waku/waku_rln_relay/rln_relay.nim

```diff
@@ -103,58 +103,54 @@ proc stop*(rlnPeer: WakuRLNRelay) {.async: (raises: [Exception]).} =
   await rlnPeer.groupManager.stop()
 
 proc hasDuplicate*(rlnPeer: WakuRLNRelay,
+                   epoch: Epoch,
                    proofMetadata: ProofMetadata): RlnRelayResult[bool] =
   ## returns true if there is another message in the  `nullifierLog` of the `rlnPeer` with the same
   ## epoch and nullifier as `proofMetadata`'s epoch and nullifier
   ## otherwise, returns false
   ## Returns an error if it cannot check for duplicates
 
-  let externalNullifier = proofMetadata.externalNullifier
   # check if the epoch exists
-  if not rlnPeer.nullifierLog.hasKey(externalNullifier):
+  let nullifier = proofMetadata.nullifier
+  if not rlnPeer.nullifierLog.hasKey(epoch):
     return ok(false)
   try:
-    if rlnPeer.nullifierLog[externalNullifier].contains(proofMetadata):
+    if rlnPeer.nullifierLog[epoch].hasKey(nullifier):
       # there is an identical record, mark it as spam
       return ok(true)
 
-    # check for a message with the same nullifier but different secret shares
-    let matched = rlnPeer.nullifierLog[externalNullifier].filterIt((
-        it.nullifier == proofMetadata.nullifier) and ((it.shareX != proofMetadata.shareX) or
-        (it.shareY != proofMetadata.shareY)))
-
-    if matched.len != 0:
-      # there is a duplicate
-      return ok(true)
-
     # there is no duplicate
     return ok(false)
 
-  except KeyError as e:
+  except KeyError:
     return err("the epoch was not found")
 
 proc updateLog*(rlnPeer: WakuRLNRelay,
+                epoch: Epoch,
                 proofMetadata: ProofMetadata): RlnRelayResult[void] =
   ## saves supplied proofMetadata `proofMetadata`
   ## in the `nullifierLog` of the `rlnPeer`
   ## Returns an error if it cannot update the log
 
-  let externalNullifier = proofMetadata.externalNullifier
-  # check if the externalNullifier exists
-  if not rlnPeer.nullifierLog.hasKey(externalNullifier):
-    rlnPeer.nullifierLog[externalNullifier] = @[proofMetadata]
+  # check if the epoch exists
+  if not rlnPeer.nullifierLog.hasKey(epoch):
```

Comment:
maybe [hasKeyOrPut proc](https://nim-lang.org/docs/tables.html#hasKeyOrPut,OrderedTable[A,B],A,B) or [mgetOrPut proc](https://nim-lang.org/docs/tables.html#mgetOrPut,OrderedTable[A,B],A,B) could simplify this and make it safer?

at your discretion

Reply:
addressed in 6d20aef7